// remx/dist/legacyRemx/connect.js
Object.defineProperty(exports,"__esModule",{value:true});exports.connect=exports.touchGlobalKey=exports.triggerStateUpdate=undefined;var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _jsxFileName='src/legacyRemx/connect.js';var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _lodash=require('lodash.isfunction');var _lodash2=_interopRequireDefault(_lodash);
var _react=require('react');var _react2=_interopRequireDefault(_react);
var _mobx=require('mobx');
var _custom=require('mobx-react/custom');
var _logger=require('./logger');var Logger=_interopRequireWildcard(_logger);function _interopRequireWildcard(obj){if(obj&&obj.__esModule){return obj;}else{var newObj={};if(obj!=null){for(var key in obj){if(Object.prototype.hasOwnProperty.call(obj,key))newObj[key]=obj[key];}}newObj.default=obj;return newObj;}}function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}

var globalObservableKey=(0,_mobx.observable)({key:1});

var triggerStateUpdate=exports.triggerStateUpdate=(0,_mobx.action)(function(){
globalObservableKey.key++;
});

var touchGlobalKey=exports.touchGlobalKey=function touchGlobalKey(){
(function(){return globalObservableKey.key;})();
};

var connect=exports.connect=function connect(mapStateToProps){
return function(Comp){
var connectedComponent=createConnectedComponent(Comp,mapStateToProps);
copyStaticMembers(connectedComponent,Comp);
return connectedComponent;
};
};
var createConnectedComponent=function createConnectedComponent(Comp,mapStateToProps){
var getMappedProps=function getMappedProps(ownProps){
if((0,_lodash2.default)(mapStateToProps)){
Logger.startLoggingMapStateToProps();
var propsFromState=mapStateToProps(ownProps);
Logger.endLoggingMapStateToProps(Comp.name,propsFromState);
return propsFromState;
}
return{};
};var

Hoc=function(_React$Component){_inherits(Hoc,_React$Component);
function Hoc(props){_classCallCheck(this,Hoc);var _this=_possibleConstructorReturn(this,(Hoc.__proto__||Object.getPrototypeOf(Hoc)).call(this,
props));

_this.originalComponentName=Comp.name;return _this;
}_createClass(Hoc,[{key:'render',value:function render()

{
return(
_react2.default.createElement(Comp,_extends({},
this.props,
getMappedProps(this.props),{
globalObservableKey:globalObservableKey.key,__source:{fileName:_jsxFileName,lineNumber:44}})));


}}]);return Hoc;}(_react2.default.Component);


return(0,_custom.observer)(Hoc);
};

var copyStaticMembers=function copyStaticMembers(target,src){
Object.keys(src).forEach(function(key){
target[key]=src[key];
});
};
